iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp

```diff
@@ -0,0 +1,357 @@
+                          SubscriberPortProducerType& subscriberPortRouDi,
+                          SubscriberPortUser& subscriberPortUser)
+    {
+        bool finished{false};
+        optional<CaproMessage> maybeCaproMessage;
+        CaproMessage caproMessage;
+
+        // Wait for publisher to be ready
+        caproMessage = waitForCaproMessage(m_concurrentCaproMessageVector, CaproMessageType::OFFER);
+
+        // Subscribe to publisher
+        subscriberPortUser.subscribe();
+        maybeCaproMessage = subscriberPortRouDi.getCaProMessage();
+        if (maybeCaproMessage.has_value())
+        {
+            caproMessage = maybeCaproMessage.value();
+            m_concurrentCaproMessageVector->push_back(caproMessage);
+        }
+
+        // Wait for subscription ACK from publisher
+        caproMessage = waitForCaproMessage(m_concurrentCaproMessageVector, CaproMessageType::ACK);
+
+        // Let RouDi change state to finish subscription
+        static_cast<void>(subscriberPortRouDi.dispatchCaProMessage(caproMessage));
+
+        // Subscription done and ready to receive samples
+        while (!finished)
+        {
+            // Try to receive chunk
+            subscriberPortUser.getChunk()
+                .and_then([&](optional<const ChunkHeader*>& maybeChunkHeader) {
+                    if (maybeChunkHeader.has_value())
+                    {
+                        auto chunkHeader = maybeChunkHeader.value();
+                        m_receiveCounter++;
+                        subscriberPortUser.releaseChunk(chunkHeader);
+                    }
+                    else
+                    {
+                        // Nothing received -> check if publisher(s) still running
+                        if (m_publisherRunFinished.load(std::memory_order_relaxed) == numberOfPublishers)
+                        {
+                            finished = true;
+                        }
+                    }
+                })
+                .or_else([](ChunkReceiveError) {
+                    // Errors shall never occur
+                    FAIL();
+                });
+        }
+    }
+
+    void publisherThread(uint32_t publisherThreadIndex,
+                         PublisherPortRouDi& publisherPortRouDi,
+                         PublisherPortUser& publisherPortUser)
+    {
+        optional<CaproMessage> maybeCaproMessage;
+        CaproMessage caproMessage;
+
+        // Publisher offers its service
+        publisherPortUser.offer();
+
+        // Let RouDi change state and send OFFER to subscriber
+        maybeCaproMessage = publisherPortRouDi.getCaProMessage();
+
+        if (publisherThreadIndex == 0)
+        {
+            // First publisher thread will sync with subscriber
+            if (maybeCaproMessage.has_value())
+            {
+                caproMessage = maybeCaproMessage.value();
+                auto guardedVector = m_concurrentCaproMessageVector.GetScopeGuard();
```

Comment:
why do you need the Scope guard here but not some lines below where you send the ACK back?

Reply:
Must be a leftover. Remove the `GetScopeGuard()` as only single access is needed here so a `->` is sufficient.